File: 22FALL/200.number-of-islands.py
#
# @lc app=leetcode id=200 lang=python3
#
# [200] Number of Islands
#

# @lc code=start
from collections import deque
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        m, n = len(grid), len(grid[0])
        eswn = [(1, 0), (0, -1), (-1, 0), (0, 1)]

        def submerge_island(x, y):
            lands = deque([(x, y)])
            while lands:
                i, j = lands.popleft()            
                for dx, dy in eswn:
                    if 0 <= i+dx < m and 0 <= j+dy < n and grid[i+dx][j+dy] == '1':
                        grid[i+dx][j+dy] = '0'
                        lands.append((i+dx, j+dy))
            return
                
        ans = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    ans += 1
                    grid[i][j] = '0'
                    submerge_island(i, j)
        return ans
        
# @lc code=end

File: VSCode_CPP/200.number-of-islands.cpp
/*
 * @lc app=leetcode id=200 lang=cpp
 *
 * [200] Number of Islands
 */

// @lc code=start
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        
        int ans = 0;
        for (int row=0; row<m; row++) {
            for (int col=0; col<n; col++) {
                if (grid[row][col] == '1') {
                    dfs(row, col, grid, m, n);
                    ans++;
                }
            }
        }
        return ans;
    }

    void dfs(int row, int col, vector<vector<char>>& grid, int m, int n) {
        grid[row][col] = '0';
        int dir[] = {-1, 0, 1, 0, -1};

        for(int i=0; i<4; i++) {
            int r = row + dir[i];
            int c = col + dir[i+1];
            if (0 <= r && r<m && 0 <= c && c < n && grid[r][c] == '1') 
                dfs(r, c, grid, m, n);
        }
    }
};
// @lc code=end

